### PR TITLE
#58 refactor into monorepo with nx

### LINKED ISSUE #58: Refactor repo as nx monorepo
Idea of this issue is to get a discussion going and decide how we might want to organise are project in a way that makes managing our front-end and multiple back-end dependencies a bit easier.

## Expected Behavior

Some of the things I think we could benefit from:

- Shared dev configs (prettier, babel, typescript, etc.)
- Shared types (e.g. shared interfaces/schemas for consistency between front-end and back-end services).
- Shared data (e.g. data for populating back-end can also be used by front-end as test fixtures).
- End-to-end (E2E) testing (e.g. test services working together to replicate functionality in the wild).
- Easier scalability (e.g. if we decided to separate the admin portal out as a separate web app, this would be made easy with shared components etc.).

## Current Behavior

Currently we manage all our projects separately. This has already been a bit tricky (e.g. setting up tests), and I think this could get a bit messy as we scale. Thinking is that we could benefit a lot from having some tools to speed things up and to keep things nice and tidy (would also be a good chance to research monorepo toolkits for omsupply).

## Possible Solution

A few possible options:

- [bazel](https://bazel.build/).
- [baur](https://github.com/simplesurance/baur).
- [buck](https://buck.build/).
- [lerna](https://lerna.js.org/).
- [nx](https://nx.dev/).
- [pants](https://www.pantsbuild.org/).

Of these, `nx` stands out to me as the most comprehensive and best documented... but feedback welcome!

### PR DESCRIPTION
**Pull request by katherine-sussol**
Fixes #58.

## Description

Initial refactor based on https://github.com/openmsupply/unified-codes/issues/58#issuecomment-666850871 
To the best of my knowledge everything is working as before (did cursory dev-test of all the services). Keen on feedback (particularly any tidy up tips or things that may not have been configured 100% correctly as the settings/tools have changed a bit). Apologies for the giant PR!

Additional Changes:
- All application dependencies removed from subfolders and merged (best attempt) into nx-workspace\package.json
- Fastify dependency aligned to 3.0.2 (API was previously on 2.14.1, data-service was on 3.0.2) - some minor namespace changes were needed)
- Several workspace generated items now live at the root of the workspace repo:
  - `libs/` (currently empty)
  - `tools/` (currently empty)
  - `dist/` (all transpiled apps go here)
  - `web-e2e` app (thinking of removing this unless others are keen to try out [cypress?)](https://www.cypress.io/)
  - nx default `README.md` (kind of helpful but thinking of maybe moving to wiki?)
  - `.eslintrc` this is default (i.e. not me trying to sneak in my own preferences)
  - `.prettierc` migrated from API project and moved to workspace config
  - `nx.json/workspace.json` - nx workspace configs
  - `tsconfig.base.json` best attempt at merging + allowJs option enabled (due to presence of js files)

**Running:**

Executed at workspace root - `unified-codes\nx-workspace\`:
- Install Dependencies    
  - `npm install`
- Build 
  - ALL: `nx run-many --target=build --all --maxParallel=3` 
  - PROJECT: `nx build [project-name]`
- Run 
  - `nx serve [project-name]`
- Start Dgraph 
  - run `/apps/data-service/src/app/database/docker.sh`
- Run Storybook
  - `nx run web:storybook`

**Linting/Formatting/Testing:**
[TARGET] = [lint|format|test]
  - ALL: `nx run-many --target=[TARGET] --all --parallel --maxParallel=3`
  - PROJECT: `nx [TARGET] [project-name]`

**Next Steps**
- Add authentication-service as separate application once #66 is merged into master

### Documentation:

Tick below if one of the following applies:

1. Documentation has been added to cover changes introduced by this PR.
2. This PR requires no changes to documentation.

- [ ] Documentation OK.

Docs probably need an update after this

### Tests:

Tick below if one of the following applies:

1. Tests have been added to cover changes introduced by this PR.
2. This PR requires no new tests.

- [x] Tests OK.

### CODE REVIEW

**Comment by mark-prins** [comment]
The `README` in the root still says to run `docker.sh` - I think this should now be `nx-workspace/apps/data-service/src/app/database/docker.sh`

**Comment by katherine-sussol** [comment]
> The `README` in the root still says to run `docker.sh` - I think this should now be `nx-workspace/apps/data-service/src/app/database/docker.sh`

Ah yes, I think some of the other file names/locations were a little outdated as well, did a quick update, though now also thinking there should be some further instructions now that we actually have an application...or a few. 

cd9c64cd05aee2d7882edf1c7d8163f9faac60c4

File: nx-workspace/package.json [hunk 1]
```diff
@@ -0,0 +1,104 @@
+{
+  "name": "monorepo",
```
[the remark this answers is not in the document]
Comment (katherine-sussol) [reply nx-workspace/package.json:2]: Whoops `monorepo` was only meant to be a placeholder. Yeah, will have to tidy that one up later. uni-something drug codes database sounds good? Too many options :(

File: nx-workspace/.gitignore [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+.project
+.classpath
+.c9/
+*.launch
+.settings/
+*.sublime-workspace
+
+# IDE - VSCode
+.vscode/*
+!.vscode/settings.json
```
[the remark this answers is not in the document]
Comment (katherine-sussol) [reply nx-workspace/.gitignore:22]: Oh - this wasn't intentional, I must have missed the generated '!' lines and didn't notice as my workspace is set up a little differently (i.e. the .vscode file sits outside a specific repo). Think we're all still in agreement to remove these from repo 😀  - can do this in a later PR.

File: nx-workspace/package.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (mark-prins) [reply nx-workspace/package.json:2]: shhh! I've just removed them from the prettier PR (#91) which I was tidying up 😉

**Pull request merged**